Feature/zenko 713 new management agent

jeremyds opened the pull request:
# Pull request template

## Description

### Motivation and context

Why is this change required? What problem does it solve?

### Related issues

Please use the following link syntaxes #600 to reference issues in the
current repository

## Checklist

### Add tests to cover the changes

New tests added or existing tests modified to cover all changes

### Code conforms with the [style guide](https://github.com/scality/Guidelines/blob/master/CONTRIBUTING.md#coding-style-guidelines)

### Sign your work

In order to contribute to the project, you must sign your work
https://github.com/scality/Guidelines/blob/master/CONTRIBUTING.md#sign-your-work

Thank you again for contributing! We will try to test and integrate the change
as soon as we can.

jeremyds committed:
ft: ZENKO-713: management agent websocket server

This commit creates a websocket server which adds a '/watch' route. The
management agent registers a listener to the new configuration event and
then sends it to its client.

bert-e commented:
# Hello jeremyds,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get
information on this process.





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @dora-korpar

* @RachedBenMustapha

* @rahulreddy

jeremyds commented:
for reviewers: only the last commit (ZENKO 713) is relevant, others are in [this PR.](https://github.com/scality/S3/pull/1378)

bert-e commented:
# Conflict




There is a conflict between your branch `feature/ZENKO-713-new-management-agent` and the
destination branch `development/8.1`.

Please resolve the conflict on **the feature branch** (`feature/ZENKO-713-new-management-agent`).

```bash
 $ git fetch
 $ git checkout origin/feature/ZENKO-713-new-management-agent
 $ git merge origin/development/8.1
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:feature/ZENKO-713-new-management-agent
```

jeremyds commented on managementAgent.js:
```diff
@@ -0,0 +1,67 @@
+    addOverlayMessageListener((remoteOverlay) => {
+        wss.clients.forEach((client) => {
+            if (client.readyState !== client.OPEN) {
+                logger.warning('client socket not in ready state', { client });
```
logger.error

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @dora-korpar

* @RachedBenMustapha

* @rahulreddy

rachedbenmustapha commented on lib/management/configuration.js:
```diff
@@ -13,6 +13,9 @@ const managementDatabaseName = 'PENSIEVE';
 const { decryptSecret } = arsenal.pensieve.credentialUtils;
 
+
+let configurationListeners = [];
```
This can be const if it's pushed into

jeremyds replied:
not relevant anymore, code has been moved to push.js

rachedbenmustapha commented on managementAgent.js:
```diff
@@ -3,8 +3,16 @@ const Uuid = require('uuid');
+
+
+//TODO: signal handler to cleanly shutdown server
+//TODO: ping messages between server and client
```
I think it would be good to implement the ping part in this PR

jeremyds replied:
done (I missed this comment)

rachedbenmustapha reviewed (changes requested):
In order to be more resilient to disconnections and timing issues, we should also send the latest overlay on the websocket as soon as a client is connected.

jeremyds committed:
ft: ZENKO-713: add callback to initManagement

The future management agent process will init the management before to
launch a websocket server. This commit adds an optional callback to the
initManagement function, call once the init is done, with the loaded
overlay as parameter.

jeremyds committed:
ft: ZENKO-713: management agent message type definition.

Define a new message type for the futur communication between the
management agent and its client.

jeremyds committed:
ft: ZENKO-713: overlay event listener

The current management code runs in S3 patches and saves received
overlay. The new management process, which will manage all the
management feature, won't patch and save the new overlay, it is a S3
specific task.

This commit adds a new overlay event listerner from the management
push code (where the new overlay is received). It will allow the
management process to send to its client the new overlay.

A new env varialble is added in this commit. It is temporary used to
remove safely the management feature from the S3 process, keeping the
previous behavior in a first time.

jeremyds committed:
ft: ZENKO-712: linter errors.

jeremyds commented on lib/management/configuration.js:
```diff
@@ -300,10 +310,22 @@ function loadCachedOverlay(log, callback) {
         });
 }
 
+/**
```
should be removed, interactive rebase error I guess, now in push.js.

jeremyds committed:
ft: ZENKO-713: log msg improvement

rachedbenmustapha commented on managementAgent.js:
```diff
@@ -55,5 +134,19 @@ function managementAgentWS() {
+    checkBrokenConnections() {
+        this.wss.clients.forEach((client) => {
+            if (!client.isAlive) {
+                logger.error('close broken connection');
```
This should be info instead of error - errors are for abnormal conditions and a client restarting is normal.

jeremyds replied:
done

rachedbenmustapha commented on lib/management/configuration.js:
```diff
@@ -256,12 +259,19 @@ function saveConfigurationVersion(cachedOverlay, remoteOverlay, log, cb) {
             {}, log, error => {
                 if (error) {
-                    log.error('could not save configuration version',
+                    log.error('could not save configuration',
```
It's not new with this change but since there's already a diff here: we should not continue to update the version pointer if there was a failure saving the overlay in the first place.

jeremyds replied:
right, I am going to fix it in a separate commit.

rachedbenmustapha commented on managementAgent.js:
```diff
@@ -1,51 +1,130 @@
-    wss.on('error', error => {
-        logger.error('management agent websocket server error', { error });
-    });
+        if (!socket.readyState === socket.OPEN) {
```
Cosmetic but it took me a few seconds to see that the condition is reversed, I suggest using `!==` instead

jeremyds replied:
right, not easy to read

rachedbenmustapha commented on managementAgent.js:
```diff
@@ -1,51 +1,130 @@
+        metadata.setup(error => {
+            if (error) {
+                const msg = `failed to setup metadata: ${error}`;
+                throw new Error(msg);
```
We should not throw in async callbacks, instead call a callback with the failure so that the toplevel code can schedule a retry

jeremyds replied:
done

rachedbenmustapha commented on managementAgent.js:
```diff
@@ -1,51 +1,130 @@
+                    this.startServer();
+                } else {
+                    const msg = `failed to init management: ${error}`;
+                    throw new Error('failed to init management');
```
Same here, we should not throw

jeremyds replied:
done too

rachedbenmustapha commented on managementAgent.js:
```diff
@@ -1,51 +1,130 @@
+    }
+
+    onNewOverlay(remoteOverlay) {
+        //XXX patch configuration?
```
I don't think the server has anything to patch as the overlay only contains s3-protocol specific items. On thing that will come later is the `browserAccess` flag

jeremyds replied:
thanks for the answer, XXX to remove

rachedbenmustapha commented on managementAgent.js:
```diff
@@ -1,51 +1,130 @@
-        logger.error('management agent websocket server error', { error });
-    });
+        if (!socket.readyState === socket.OPEN) {
+            logger.error('client socket not in ready state', { socket });
```
This could include a lot of unpleasant data to look at that can be big or circular, so for clarity and for preserving performance we should select a few meaningful attributes (if available) such as `host`, `port`, `readyState` and so on

jeremyds replied:
done

rachedbenmustapha commented on lib/management/agentClient.js:
```diff
@@ -0,0 +1,16 @@
+    /** Message that contains a configuration overlay */
+    NEW_OVERLAY_VERSION: 1,
+    /** Message that contains the loaded overlay */
+    LOADED_OVERLAY: 2,
```
Why is there a different message type `LOADED_OVERLAY`? I think only one message type to push overlays would be clearer.

jeremyds replied:
I thought it maybe useful to distinguish the loaded overlay from a new one. But as long as I don't have a precise use case in mind, let's have only one message type for the moment.

jeremyds committed:
ft: ZENKO-713: fix management::SaveconfigurationVersion error handling

rachedbenmustapha commented on managementAgent.js:
```diff
@@ -1,59 +1,159 @@
+        metadata.setup(error => {
+            if (error) {
+                const msg = `failed to setup metadata: ${error}`;
+                return cb(new Error(msg));
```
Real `Error` objects usually don't print well in our json logs so just forwarding the error we get to the callback is safer. A log before that saying `log.error('failed to setup metadata', { error })` would be enough to track

jeremyds replied:
done

rachedbenmustapha commented on lib/management/push.js:
```diff
@@ -141,10 +146,7 @@ function startWSManagementClient(pushEndpoint, instanceId, token) {
-                host: host,
-                port: port,
-            });
+            socket = net.createConnection(host, port);
```
The order is wrong here, its port then host (nodejs nicety). Why the change?

jeremyds replied:
wow... linter complains about the object form {port: value, host: value}

rachedbenmustapha replied:
Maybe it wants `{ host, port }` since the vars have the same name as the keys

nicolas2bert commented on lib/management/index.js:
```diff
@@ -80,14 +82,14 @@ function initManagement(log) {
             }
-            return patchConfiguration(conf, log,
-                err => cb(err, instanceId, token));
+            return patchConfiguration(overlay, log,
```
It might be clearer to add another waterfall callback here for `patchConfiguration()`. 
`(overlay, instanceId, ... cb) => patchConfiguration(overlay...)`

jeremyds replied:
right, done

nicolas2bert commented on managementAgent.js:
```diff
@@ -1,59 +1,159 @@
+
+class ManagementAgentServer {
+    constructor() {
+        this.port = _config.managementAgent.port || 8010;
```
Not sure how `_config.managementAgent.port` can be `undefined` or `0` since you already asserted it in the `Config.js`. _Better safe than sorry?_

jeremyds replied:
yes, exactly

nicolas2bert commented on managementAgent.js:
```diff
@@ -1,59 +1,159 @@
+        this.port = _config.managementAgent.port || 8010;
+        this.wss = null;
+        this.loadedOverlay = null;
+
```
You could also have done it this way:
```
this.stop = this.stop.bind(this);

process.on('SIGINT', this.stop);

process.on('SIGHUP', this.stop);

process.on('SIGQUIT', this.stop);

process.on('SIGTERM', this.stop);
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:



* 2 peers





Peer approvals *must* include at least 1 approval from the following list:

* @rahulreddy

* @dora-korpar

* @RachedBenMustapha

rachedbenmustapha commented on managementAgent.js:
```diff
@@ -1,59 +1,169 @@
 }
+
+const server = new ManagementAgentServer();
+server.start();
```
If `ManagementAgentServer.start()` fails and we don't provide a callback that exits, are we sure that the process will exit with an error status? When deployed in k8s for example if mongodb is not yet available we want to be sure that we retry initializing later. Arguably we could crash as well but it only works well when deployed inside an orchestrator.

jeremyds committed:
ft: ZENKO-713: metadata.setup in initManagement.

The future management process needs metadata setup. Call added at
initManagement function beginning, conditioned to the new management
agent use. setup the metadata in this function allows to take advantage
of its auto retry mechanism.

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.1`



The following branches will **NOT** be impacted:

* `development/6.4`
* `development/7.4`
* `development/8.0`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of Release Engineering will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of Release Engineering now.

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.1`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.4`
* `development/8.0`



Please check the status of the associated issue ZENKO-713.

Goodbye jeremyds.
